### PR TITLE
improve block finding algorithm yet again

### PR DESCRIPTION
**Pull request by pfitzseb**
Fixes https://discourse.julialang.org/t/strange-behavior-in-live-line-evaluation/32497 and makes a few general improvements.

### CODE REVIEW

**Comment by aviatesk** [comment]
will try to take a look at this (I'm not sure I can understand that correctly)

File: lib/misc/blocks.js [hunk 1]
```diff
@@ -46,13 +46,12 @@ function walkForward (ed, start) {
...
+
     if (isStart(lineInfo)) {
       break
     }
     if (isEnd(lineInfo)) {
-      if (!(forLines(ed, start, mark-1).length === 0)) {
-        end = mark
-      }
+      end = mark
     } else if (!(isBlank(lineInfo) || isStart(lineInfo))) {
```
Comment (aviatesk) [inline lib/misc/blocks.js]: nit: we can remove the second `isStart` condition since it's always `false` in this path.

File: lib/misc/blocks.js [hunk 2]
```diff
@@ -16,21 +16,21 @@ function isBlank ({line, scope}, allowDocstrings = false) {
...
 }
 function isCont ({ line }) {
   return line.match(/^(else|elseif|catch|finally)\b/)
 }
 function isStart (lineInfo) {
-  for (const s of lineInfo.scope) {
-    if (/\bstring\b/.test(s)) {
-      return false
-    }
-  }
```
Comment (aviatesk) [inline lib/misc/blocks.js]: um, so if we remove this condition check, the code like below won't get selected as expected again:
```jl
ret = """
multiline string
...
""" |> somefunc
```

But I've not been able to figure out how to handle this while correctly run pieces like

```jl
str = "a line containing string scope"
str2 = "another line containing string scope"
...
```

Comment (pfitzseb) [reply lib/misc/blocks.js]: Urgh, true. I only tested
```
ret = """
  multiline string
  aksjdha
""" |> somefunc
```
Will take another look...

Comment (aviatesk) [reply lib/misc/blocks.js]: (I'm trying not to say let's use CSTParser for this too ...)

Comment (pfitzseb) [reply lib/misc/blocks.js]: Those cases work now.

**Comment by aviatesk** [comment]
Just done manual tests (didn't look at code) but found this PR solves every case that is mentioned here !

**Comment by pfitzseb** [comment]
Nice, let's go with this then. I'm sure someone will complain if I broke something else :P

**Pull request merged**